walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,111 @@
+    List<TimeRange> availableTime = getAvailableTime(unavailableTime, request.getDuration());
+    return availableTime;
+  }
+
+  /** Get a collection of all attendees' unavailable time ranges (due to other events). */
+  private List<TimeRange> getAttendeesUnavailableTime(
+      Collection<Event> events, Collection<String> meetingAttendees) {
+    List<TimeRange> unavailableTime = new ArrayList<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      // If any attendee in the meeting joins that event as well, the time range of the event is
+      // added to the unavailableTime collection.
+      if (!Collections.disjoint(meetingAttendees, eventAttendees)) {
+        unavailableTime.add(event.getWhen());
+      }
+    }
+    return unavailableTime;
+  }
+
+  /** Return a collection of time ranges in the time order without any overlap among each other. */
+  private List<TimeRange> getNoOverlapTime(List<TimeRange> unavailableTime) {
+    List<TimeRange> noOverlapTime = new ArrayList<>();
+    for (TimeRange newTime : unavailableTime) {
+      if (noOverlapTime.isEmpty()) {
+        noOverlapTime.add(newTime);
+        continue;
+      }
+      // Compare and combine the new time with any existing time range in collection in order.
+      List<TimeRange> localNoOverlapTime = new ArrayList<>();
+      for (int i = 0; i < noOverlapTime.size(); i++) {
+        TimeRange existingTime = noOverlapTime.get(i);
+        // Case 1. Existing time is earlier than new time
+        if (existingTime.end() <= newTime.start()) {
+          localNoOverlapTime.add(existingTime);
+        } else if (existingTime.start() >= newTime.end()) {
+          // Case 2. New time is earlier than existing time
+          localNoOverlapTime.add(newTime);
+          localNoOverlapTime.addAll(noOverlapTime.subList(i, noOverlapTime.size()));
+          break;
+        } else if (existingTime.contains(newTime)) {
+          // Case 3. Existing time contains new time
+          localNoOverlapTime.addAll(noOverlapTime.subList(i, noOverlapTime.size()));
+          break;
+        } else if (newTime.contains(existingTime)) {
+          // Case 4. New time contains existing time
+        } else {
+          // Case 5. Overlap (no contain) exists between existing time and new time
+          newTime = TimeRange.fromStartEnd(Math.min(newTime.start(), existingTime.start()),
+              Math.max(newTime.end(), existingTime.end()), false);
+        }
+      }
+      // Check whether newTime has never been included because it is the last one.
+      if (localNoOverlapTime.isEmpty()) {
+        localNoOverlapTime.add(newTime);
+      } else {
+        if (localNoOverlapTime.get(localNoOverlapTime.size() - 1).end() < newTime.end()) {
+          localNoOverlapTime.add(newTime);
+        }
+      }
+      noOverlapTime = localNoOverlapTime;
+    }
+    return noOverlapTime;
```

Comment:
This algorithm seems to be O(n^2). There might be a more efficient algorithm. This is not really important for this codelab.

Reply:
Yes. I also think it is O(n^2). I will consider the algorithm. If I have any better idea, I may add it as TODO.

Reply:
Yes. I considered one way to have less time complexity. Instead of doing both sorting and merging time slots (i.e., eliminating overlaps) in this one function, I can do sort first and then separately eliminate overlaps for the sorted time range. The time complexity of sorting will be smaller than O(n^2), and the time complexity of eliminating overlaps will only be O(n).
While because it does change many codes for the current algorithm, I do not plan to change it here, as it is not really the target for this exercise as you said. I will have a TODO notice separately for this possible update.